src/windows.rs

```diff
@@ -0,0 +1,480 @@
+        let normalised_current = (self.current - self.min) as f64;
+        (normalised_current / normalised_max * 100.0).round() as u32
+    }
+
+    fn percentage_to_current(&self, percentage: u32) -> u32 {
+        let normalised_max = (self.max - self.min) as f64;
+        let fraction = percentage as f64 / 100.0;
+        let normalised_current = fraction * normalised_max;
+        normalised_current.round() as u32 + self.min
+    }
+}
+
+fn ddcci_get_monitor_brightness(device: &Brightness) -> Result<DdcciBrightnessValues, SysError> {
+    unsafe {
+        let mut v = DdcciBrightnessValues::default();
+        BOOL(GetMonitorBrightness(
+            device.physical_monitor.0,
+            &mut v.min,
+            &mut v.current,
+            &mut v.max,
+        ))
+        .ok()
+        .map(|_| v)
+        .map_err(|e| SysError::GettingBrightnessFailed {
+            device_name: device.device_name.clone(),
+            function_name: "GetMonitorBrightness",
+            last_error: e.code(),
+        })
+    }
+}
+
+fn ddcci_set_monitor_brightness(device: &Brightness, value: u32) -> Result<(), SysError> {
+    unsafe {
+        BOOL(SetMonitorBrightness(device.physical_monitor.0, value))
+            .ok()
+            .map_err(|e| SysError::SettingBrightnessFailed {
+                device_name: device.device_name.clone(),
+                function_name: "SetMonitorBrightness",
+                last_error: e.code(),
+            })
+    }
+}
+
+#[derive(Debug)]
+// "Each level is a value from 0 to 100"
+struct IoctlSupportedBrightnessLevels(Vec<u8>);
+
+impl IoctlSupportedBrightnessLevels {
+    fn get_nearest(&self, percentage: u32) -> u8 {
+        self.0
+            .iter()
+            .min_by_key(|&num| (*num as i64 - percentage as i64).abs())
+            .map(|x| *x)
+            .unwrap_or(0)
+    }
+}
+
+fn ioctl_query_supported_brightness(
+    device: &Brightness,
+) -> Result<IoctlSupportedBrightnessLevels, HRESULT> {
+    unsafe {
+        let mut bytes_returned = 0;
+        let mut out_buffer = Vec::<u8>::with_capacity(256);
+        DeviceIoControl(
+            device.file_handle.0,
+            IOCTL_VIDEO_QUERY_SUPPORTED_BRIGHTNESS,
+            std::ptr::null_mut(),
+            0,
+            out_buffer.as_mut_ptr() as *mut c_void,
+            out_buffer.capacity() as u32,
+            &mut bytes_returned,
+            std::ptr::null_mut(),
+        )
+        .ok()
+        .map(|_| {
+            out_buffer.set_len(bytes_returned as usize);
+            IoctlSupportedBrightnessLevels(out_buffer)
+        })
+        .map_err(|e| e.code())
+    }
+}
+
+fn ioctl_query_display_brightness(device: &Brightness) -> Result<u32, SysError> {
+    unsafe {
+        let mut bytes_returned = 0;
+        let mut display_brightness = DISPLAY_BRIGHTNESS::default();
+        DeviceIoControl(
+            device.file_handle.0,
+            IOCTL_VIDEO_QUERY_DISPLAY_BRIGHTNESS,
```

Comment:
[The documentation](https://docs.microsoft.com/en-us/windows/win32/power/ioctl-video-query-display-brightness) states:
> Support for this control code was removed in Windows Server 2008 and Windows Vista.
> End of client support: Windows XP with SP2

It seems to work on my Windows 10 laptop though. I wonder if it may not work with other hardware. AFAICT the recommended solution needs COM which brings its own headaches.

Reply:
Yeah that is really odd and annoying, I didn't actually see that when writing it, since I mainly read the IOCTL_VIDEO_QUERY_SUPPORTED_BRIGHTNESS and IOCTL_VIDEO_SET_DISPLAY_BRIGHTNESS docs which don't mention the deprecation, it seem really strange to only deprecate one of the 3 functions in the API?!

It works fine on my laptop (Windows 10 20H2), and I personally have no desire to rewrite it for the time being, but if you or someone else in future wants to change to the WMI APIs then go ahead.

Reply:
> [...] it seem really strange to only deprecate one of the 3 functions in the API?!

Agreed.